lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3860,21 +3860,130 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
   case Intrinsic::usub_sat:
   case Intrinsic::sadd_sat:
   case Intrinsic::ssub_sat: {
-    SPIRVWord ExtOp;
-    if (IID == Intrinsic::uadd_sat)
-      ExtOp = OpenCLLIB::UAdd_sat;
-    else if (IID == Intrinsic::usub_sat)
-      ExtOp = OpenCLLIB::USub_sat;
-    else if (IID == Intrinsic::sadd_sat)
-      ExtOp = OpenCLLIB::SAdd_sat;
-    else
-      ExtOp = OpenCLLIB::SSub_sat;
+    if (BM->shouldUseOpenCLExtInstructionsForLLVMIntrinsic()) {
+      SPIRVWord ExtOp;
+      if (IID == Intrinsic::uadd_sat)
+        ExtOp = OpenCLLIB::UAdd_sat;
+      else if (IID == Intrinsic::usub_sat)
+        ExtOp = OpenCLLIB::USub_sat;
+      else if (IID == Intrinsic::sadd_sat)
+        ExtOp = OpenCLLIB::SAdd_sat;
+      else
+        ExtOp = OpenCLLIB::SSub_sat;
 
-    SPIRVType *Ty = transType(II->getType());
-    std::vector<SPIRVValue *> Operands = {transValue(II->getArgOperand(0), BB),
-                                          transValue(II->getArgOperand(1), BB)};
-    return BM->addExtInst(Ty, BM->getExtInstSetId(SPIRVEIS_OpenCL), ExtOp,
-                          std::move(Operands), BB);
+      SPIRVType *Ty = transType(II->getType());
+      std::vector<SPIRVValue *> Operands = {
+          transValue(II->getArgOperand(0), BB),
+          transValue(II->getArgOperand(1), BB)};
+      return BM->addExtInst(Ty, BM->getExtInstSetId(SPIRVEIS_OpenCL), ExtOp,
+                            std::move(Operands), BB);
+    }
+
+    Type *LLVMTy = II->getType();
+    SPIRVType *Ty = transType(LLVMTy);
+    Type *BoolTy = IntegerType::getInt1Ty(M->getContext());
+    if (auto *VecTy = dyn_cast<VectorType>(LLVMTy))
+      BoolTy = VectorType::get(BoolTy, VecTy->getElementCount());
+    SPIRVType *SPVBoolTy = transType(BoolTy);
+    SPIRVValue *FirstArgVal = transValue(II->getArgOperand(0), BB);
+    SPIRVValue *SecondArgVal = transValue(II->getArgOperand(1), BB);
+
+    if (IID == Intrinsic::uadd_sat) {
+      // uadd.sat(a, b) -> res = a + b, (res > a) ? res : MAX
```

Comment:
Is `res` will anyway be bigger or equal than `a`? I thought maybe the operation is between 2 unsigned integers

Reply:
If we have an integer overflow - then no. For unsigned integers it's the defined behavior.

Reply:
I understood the intrinsic description differently, here is info from LLVM lang ref (https://llvm.org/docs/LangRef.html#llvm-uadd-sat-intrinsics):

> The maximum value this operation can clamp to is the largest unsigned value representable by the bit width of the arguments. Because this is an unsigned operation, the result will never saturate towards zero.
`%res = call i4 @llvm.uadd.sat.i4(i4 8, i4 8)  ; %res = 15`

Reply:
Okay, it seems like I misunderstood the question. I answer like it was a general question, but it's related to the intrinsic. Intrinsic acts like it's described in the comment: picking between 'res' and 'MAX' depending if overflow has happen. So emulation should also pick between them. But detecting overflow is being done by the check if `a + b > a`, which can only happen during overflow.

Reply:
Aahh, I see, oops, the code and the comment do a bit an opposite.

Reply:
In cases where b = 0; res > a will return false and hence MAX will be selected. I think that is not the correct answer. I think the checking condition needs to be res >= a.
Please revisit.

Thanks

Reply:
It should be `res < a` as overflow condition.

Reply:
I was a bit confused answering, but basically @asudarsa is right, thanks!